relay-server/src/metrics_extraction/event.rs

```diff
@@ -1,109 +1,66 @@
-                    "span.group",
-                    "span.module",
-                    "span.op",
-                    "span.status_code",
-                    "span.status",
-                    "span.system",
-                    "transaction.method",
-                    "transaction.op",
-                ]
-                .map(|key| TagSpec {
-                    key: key.into(),
-                    field: Some(format!("span.data.{}", key.replace('.', "\\."))),
-                    value: None,
-                    condition: None,
-                })
-                .into(),
-            },
-            TagMapping {
-                metrics: vec![LazyGlob::new("d:spans/exclusive_time*@millisecond".into())],
-                tags: ["release", "device.class"] // TODO: sentry PR for static strings
-                    .map(|key| TagSpec {
-                        key: key.into(),
-                        field: Some(format!("span.data.{}", key.replace('.', "\\."))),
-                        value: None,
-                        condition: Some(RuleCondition::Eq(EqCondition {
-                            name: "span.data.mobile".into(),
-                            value: Value::Bool(true),
-                            options: Default::default(),
-                        })),
-                    })
-                    .into(),
-            },
-        ],
-        _conditional_tags_extended: false,
+impl Extractable for Event {
+    fn category(&self) -> DataCategory {
+        // Obtain the event's data category, but treat default events as error events for the
+        // purpose of metric tagging.
+        match DataCategory::from(self.ty.value().copied().unwrap_or_default()) {
+            DataCategory::Default => DataCategory::Error,
+            category => category,
+        }
     }
-});
 
-/// Extracts metrics from the spans of the given transaction.
-pub fn extract_metrics(event: &Event) -> Vec<Metric> {
-    relay_statsd::metric!(timer(RelayTimers::EventProcessingSpanMetricsExtraction), {
-        let mut metrics = Vec::new();
+    fn timestamp(&self) -> Option<UnixTimestamp> {
+        self.timestamp
+            .value()
+            .and_then(|ts| UnixTimestamp::from_datetime(ts.0))
+    }
+}
+
+impl Extractable for Span {
+    fn category(&self) -> DataCategory {
+        DataCategory::Span
+    }
 
+    fn timestamp(&self) -> Option<UnixTimestamp> {
+        self.timestamp
+            .value()
+            .and_then(|ts| UnixTimestamp::from_datetime(ts.0))
+    }
+}
+
+/// Extract metrics from an [`Event`].
+///
+/// The event must have a valid timestamp; if the timestamp is missing or invalid, no metrics are
+/// extracted. Timestamp and clock drift correction should occur before metrics extraction to ensure
+/// valid timestamps.
+///
+/// If this is a transaction event with spans, metrics will also be extracted from the spans.
+pub fn extract_metrics(event: &Event, config: &MetricExtractionConfig) -> Vec<Metric> {
```

Comment:
This function and module is now responsible for extracting metrics from all event types. It mostly delegates to generic functionality, but also handles spans. That's also why the implementation of the `Extractable` trait has moved here. 

The `generic` module is now fully generic.